dsh/rseq.1
.\" $Id: rseq.1,v 1.18 2007/01/10 20:36:11 garbled Exp $
.\" Copyright (c) 1998, 1999, 2000
.\"	Tim Rightnour.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by Tim Rightnour.
.\" 4. The name of Tim Rightnour may not be used to endorse or promote 
.\"    products derived from this software without specific prior written 
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd January 9, 2007
.Dt SEQ 1
.Sh NAME
.Nm rseq
.Nd run a command on a cluster in sequence
.Sh SYNOPSIS
.Nm
.Op Fl aeiqtv
.Op Fl g Ar rungroup1,...,rungroupN
.Op Fl l Ar username
.Op Fl o Ar porttimeout
.Op Fl p Ar portnum
.Op Fl w Ar node1,...,nodeN
.Op Fl x Ar node1,...,nodeN
.Op Ar command ...
.Sh DESCRIPTION
The 
.Nm
utility can be used to run a command, or group of commands on a cluster of 
machines.  The command or commands are run on a single node in the cluster,
which is incremented in sequence on each invocation of
.Nm
The following options are available:
.Bl -tag -width www
.It Fl a
The
.Fl a
option modifies the behavior of
.Nm
when dealing with stdin.  Normally 
all commands from stdin are run on the same machine.  Specifying the
.Fl a
option causes each command to be run on the next node in the seqence.
.It Fl e
Unless the
.Fl e
option is specified, stderr from remote commands will not be reported
to the user.
.It Fl i
The
.Fl i
option will list information about the current cluster, and command groupings.
It will also show you which command you are about to run, and your username if
specified with the
.Fl l
option.
.It Fl q
The
.Fl q
option does not issue any commands, but displays information about the cluster.
.It Fl v
Prints the version of ClusterIt to the stdout, and exits.
.It Fl g
If the
.Fl g
option is specified, followed by a comma separated list of group
names, the command will only be run on that group of nodes.  A node
may be a part of more than one group if desired, however running
without the
.Fl g
option will run the command on the same node as many times as it
appears in the file specified by the
.Ev CLUSTER
environment variable.  This option is silently ignored if used with the
.Fl w
option.
.It Fl l
If the
.Fl l
option is specified, followed by a username, the commands will be run 
under that userid on the remote machines.  Consideration must be taken 
for proper authentication, for this to work.
.It Fl o
The
.Fl o
option is used to set the timeout in seconds to be used when testing remote
connections.  The default is five seconds.
.It Fl p
The
.Fl p
option can be used to set the port number that testing should occur on when
testing remote connections.  The default behavior is to guess based on the
remote command name.
.It Fl t
The
.Fl t
option causes
.Nm
to attempt a connection test to each node prior to attempting to run the
remote command.  If the test fails for any reason, the remote command will
not be attempted.  This can be useful when clusterfiles have suffered bitrot
and some nodes no longer exist, or might be down for maintenance.  The default
timeout is 5 seconds.  The timeout can be changed with the
.Fl o
option.
.Nm
will attempt to guess the port number of the remote service based on your
.Ev RCMD_CMD
setting.  It knows about ssh and rsh.  If
.Nm
fails to guess your port correctly, you may use the
.Fl p
argument to set the remote port number.  If the
.Ev RCMD_TEST
environment variable exists, the testing will automatically take place.
.It Fl w
If the
.Fl w
option is specified, followed by a comma delimited list of machine names,
the
.Ar command
will be run on each node in the list.  Without this option,
.Nm
runs on the nodes listed in the file pointed to by the
.Ev CLUSTER
environment variable.
.It Fl x
The
.Fl x
option can be used to exclude specific nodes from the cluster.  The
format is the same as the
.Fl w
option, a comma delimited list of machine names.  This option is
silently ignored if used with the
.Fl w
option.
.El
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables.
.Bl -tag -width "RCMD_CMD_ARGS"
.It Ev CLUSTER
Contains a filename, which is a newline separated list of nodes
in the cluster.
.It Ev RCMD_CMD
Command to use to connect to remote machines.  The command chosen must
be able to connect with no password to the remote host.  Defaults to
.Ic rsh
.It Ev RCMD_CMD_ARGS
Arguments to pass to the remote shell command.  Defaults to none.
.It Ev RCMD_USER
The username to connect to remote machines as by default.
.It Ev SEQ_FILE
Specifies a specific sequence file.  This is a temporary file, which 
saves the number of the last node run on.  It defaults to 
.Sq /tmp/ Ns Bo pid Bc Ns .seq .
Where
.Bq pid
is the pid of the parent process of 
.Nm seq Ns .
It may be useful to wire the
.Ev SEQ_FILE
to a particular file, in order to keep proper sequence when invoking
.Nm
from subshells or scripts.
.It Ev RCMD_PORT
The port number used to test remote connections.  See the
.Fl p
flag.
.It Ev RCMD_TEST
When set,
.Nm
will automatically test all hosts before launching the remote command. See the
.Fl t
option for more information.
.It Ev RCMD_TEST_TIMEOUT
The timeout in seconds to use when testing for remote connections.
.El
.\" .Sh FILES
.Sh DIAGNOSTICS
Exit status is 0 on success, 1 if an error occurs.
.Sh SEE ALSO
.Xr dsh 1 ,
.Xr rsh 1 ,
.Xr kerberos 3 ,
.Xr hosts.equiv 5 ,
.Xr rhosts 5
.Sh HISTORY
The
.Nm
command appeared in clusterit 1.1.
.Sh AUTHOR
.Nm Seq
was written by Tim Rightnour.
.Sh BUGS
The
.Nm
command works fine when used with the
.Fl w
and
.Fl x
flags.  However the user may find that if the arguments are changed, or
the flags are omitted on a subsequent run, the next node picked may not be
predictable.  The sequence is calculated on each invocation of
.Nm
and the sequence numbers recorded in the temporary file may not match up
to a new sequence placed on the command line.


CHANGES
$Id: CHANGES,v 1.18 2007/01/04 18:59:06 garbled Exp $

Add -s option to dsh.  This adds the ability to copy a script to a remote
machine, execute it, and delete it, all in a single operation.

Add RCP_CMD_ARGS processing to pcp.

* Release 2.4 *

Fix jsd to work properly on Linux.

Fix stdin command mode on all the utilities.

Make jsh and jsd more robust.  Keep jsd from permanently losing it's nodes.

Fix pdf to work on Linux.

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *


dsh/rseq.c
/* $Id: rseq.c,v 1.23 2007/01/10 20:36:11 garbled Exp $ */
/*
 * Copyright (c) 1998, 1999, 2000
 *	Tim Rightnour.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Tim Rightnour.
 * 4. The name of Tim Rightnour may not be used to endorse or promote 
 *    products derived from this software without specific prior written 
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/wait.h>
#include <poll.h>
#include <signal.h>
#include <libgen.h>

#include "../common/common.h"

#if !defined(lint) && defined(__NetBSD__)
__COPYRIGHT(
"@(#) Copyright (c) 1998, 1999, 2000\n\
        Tim Rightnour.  All rights reserved\n");
__RCSID("$Id: rseq.c,v 1.23 2007/01/10 20:36:11 garbled Exp $");
#endif

/* externs */
extern int errno;

void do_command(char **argv, int allrun, char *username);
node_t *check_seq(void);
void sig_handler(int i);

/* globals */

int debug, errorflag, exclusion, grouping;
int seqnumber, nrofrungroups;
int rshport, testflag, porttimeout;
char **rungroup;
char **lumplist;
node_t *nodelink;
group_t *grouplist;
char *progname;

/* 
 *  seq is a cluster management tool based upon the IBM tool dsh.
 *  Seq is used to run a command on a single node, and then run the next
 *  command on the next node in the clusterfile sequentially.
 */

int main(int argc, char **argv) 
{
    extern char *optarg;
    extern int optind;
    extern char *version;

    int someflag, ch, i, allflag, showflag;
    char *p, *group, *nodename, *username;
    char **exclude;

    someflag = 0;
    seqnumber = -1;
    showflag = 0;
    exclusion = 0;
    debug = 0;
    errorflag = 0;
    allflag = 0;
    testflag = 0;
    grouping = 0;
    nrofrungroups = 0;
    porttimeout = 5; /* 5 seconds to port timeout */
    username = NULL;
    nodename = NULL;
    group = NULL;
    exclude = NULL;

    rungroup = calloc(GROUP_MALLOC, sizeof(char **));
    if (rungroup == NULL)
	bailout();

    progname = strdup(basename(argv[0]));
    
#if defined(__linux__)
    while ((ch = getopt(argc, argv, "+?adeiqtg:l:o:p:w:x:")) != -1)
#else
    while ((ch = getopt(argc, argv, "?adeiqtvg:l:o:p:w:x:")) != -1)
#endif
	switch (ch) {
	case 'a':		/* set the allrun flag */
	    allflag = 1;
	    break;
	case 'd':       /* set the debug flag */
	    debug = 1;
	    break;
	case 'e':		/* we want stderr to be printed */
	    errorflag = 1;
	    break;
	case 'i':		/* we want tons of extra info */
	    debug = 1;
	    break;
	case 'l':		/* invoke me as some other user */
	    username = strdup(optarg);
	    break;
	case 't':           /* test the nodes before connecting */
	    testflag = 1;
	    break;
	case 'p':           /* what is the rsh port number? */
	    rshport = atoi(optarg);
	    break;
	case 'q':		/* just show me some info and quit */
	    showflag = 1;
	    break;
	case 'o':               /* set the test timeout in seconds */
	    porttimeout = atoi(optarg);
	    break;
	case 'g':		/* pick a group to run on */
	    grouping = 1;
	    nrofrungroups = parse_gopt(optarg);
	    break;			
	case 'x':		/* exclude nodes, w overrides this */
	    exclusion = 1;
	    exclude = parse_xopt(optarg);
	    break;
	case 'w':		/* perform operation on these nodes */
	    someflag = 1;
	    i = 0;
	    for (p = optarg; p != NULL; ) {
		nodename = (char *)strsep(&p, ",");
		if (nodename != NULL)
		    (void)nodealloc(nodename);
	    }
	    break;
	case 'v':
	    (void)printf("%s: %s\n", progname, version);
	    exit(EXIT_SUCCESS);
	    break;
	case '?':		/* you blew it */
	    (void)fprintf(stderr,
		"usage: %s [-aeiqtv] [-p portnum] [-o timeout] "
		"[-g rungroup1,...,rungroupN]\n"
		"\t[-l username] [-x node1,...,nodeN] [-w node1,..,nodeN] "
		"[command ...]\n", progname);
	    exit(EXIT_FAILURE);
	    break;
	default:
	    break;
	}

    if (username == NULL && getenv("RCMD_USER"))
	username = strdup(getenv("RCMD_USER"));
    if (!rshport && getenv("RCMD_PORT"))
	rshport = atoi(getenv("RCMD_PORT"));
    if (!testflag && getenv("RCMD_TEST"))
	testflag = 1;
    if (porttimeout == 5 && getenv("RCMD_TEST_TIMEOUT"))
	porttimeout = atoi(getenv("RCMD_TEST_TIMEOUT"));

    rshport = get_rshport(testflag, rshport, "RCMD_CMD");

    if (!someflag)
	parse_cluster(exclude);	
    argc -= optind;
    argv += optind;
    if (showflag) {
	do_showcluster(DEFAULT_FANOUT);
	exit(EXIT_SUCCESS);
    }
    do_command(argv, allflag, username);
    exit(EXIT_SUCCESS);
}

/* this should be atomic, but *hello* this is *userland* */

static void
test_and_set(void)
{
    int i;
    char *p, *seqfile;
    char buf[MAXBUF];
    FILE *sd;
    node_t *nodeptr, *nodex;

    p = buf;
    i = 0;

    seqfile = getenv("SEQ_FILE");
    if (seqfile == NULL) {
	(void)snprintf(buf, MAXBUF, "/tmp/%d.%s", (int)getppid(), progname);
	seqfile = strdup(buf);
    }
    sd = fopen(seqfile, "r");
    if (sd == NULL)
	sd = fopen(seqfile, "w");
    else {
	fscanf(sd, "%s", p);
	seqnumber = atoi(p);
	fclose(sd);
	sd = fopen(seqfile, "w");
    }
    if (sd == NULL)
	bailout();
    nodeptr = check_seq();
    for (nodex = nodelink; nodex != nodeptr; nodex = nodex->next)
	i++;
    (void)fprintf(sd, "%d", i);
    fclose(sd);
    seqnumber = i;
}

/* return the node number of the next node in the seqence. */

node_t *
check_seq(void)
{
    int i, g;
    node_t *nodeptr;

    if (seqnumber == -1)
	return(nodelink);

    g = seqnumber;
    i = 0;
    for (nodeptr = nodelink; (nodeptr && i < g);
	 nodeptr = nodeptr->next)
	i++;
    if (nodeptr == NULL || nodeptr->next == NULL)
	return(nodelink);
    return(nodeptr->next);
}


/* 
 * Do the actual dirty work of the program, now that the arguments
 * have all been parsed out.
 */

void 
do_command(char **argv, int allrun, char *username)
{
    struct sigaction signaler;
    FILE *fd, *fda, *in;
    char buf[MAXBUF], cbuf[MAXBUF], pipebuf[2048];
    int status, piping, pollret, i, nrofargs, arg;
    size_t maxnodelen;
    char *p, *command, **rsh, *cd, **q, **cmd, *rshstring;
    node_t *nodeptr;
    struct pollfd fds[2];

    maxnodelen = 0;
    piping = 0;
    in = NULL;

    if (debug && username != NULL)
	(void)printf("As User: %s\n", username);

    /* construct the command from the remains of argv */
    for (i=0, p=*argv, q=argv; p != NULL; p = *++q)
	i += (strlen(p)+1);
    command = (char *)calloc(i+1, sizeof(char));
    for (p = *argv; p != NULL; p = *++argv ) {
	strcat(command, p);
	strcat(command, " ");
    }
    if (debug)
	(void)printf("Do Command: %s\n", command);

    for (nodeptr = nodelink; nodeptr != NULL; nodeptr = nodeptr->next)
	if (strlen(nodeptr->name) > maxnodelen)
	    maxnodelen = strlen(nodeptr->name);

    if (strcmp(command,"") == 0) {
	piping = 1;
	if (isatty(STDIN_FILENO) && piping)
	    /* are we a terminal?  then go interactive! */
	    (void)printf("%s>", progname);
	in = fdopen(STDIN_FILENO, "r");
	command = fgets(cbuf, sizeof(cbuf), in);
	/* start reading stuff from stdin and process */
	if (command != NULL)
	    if (strcmp(command,"\n") == 0)
		command = NULL;
    }

    rsh = parse_rcmd("RCMD_CMD", "RCMD_CMD_ARGS", &nrofargs);
    rshstring = build_rshstring(rsh, nrofargs);
    signaler.sa_handler = sig_handler;
    signaler.sa_flags = SA_RESTART;
    sigaction(SIGALRM, &signaler, NULL);

    if (allrun) {
	test_and_set();
	nodeptr = check_seq();
	if (testflag && rshport > 0 && porttimeout > 0) {
	    while (!test_node_connection(rshport, porttimeout, nodeptr)) {
		    fprintf(stderr, "Skipping down node %s.\n", nodeptr->name);
		    test_and_set();
		    nodeptr = check_seq();
	    }
        }
    }

    while (command != NULL) {
	if (!allrun) {
	    test_and_set();
	    nodeptr = check_seq();
	    if (testflag && rshport > 0 && porttimeout > 0) {
		    while (!test_node_connection(rshport, porttimeout,
			    nodeptr)) {
			    fprintf(stderr, "Skipping down node %s.\n",
				nodeptr->name);
			    test_and_set();
			    nodeptr = check_seq();
		    }
	    }
	}
	i = seqnumber;
	if (debug)
	    (void)printf("On node: %s\n", nodeptr->name);
	pipe(nodeptr->out.fds);
	/* we set up pipes for each node, to prepare for the
	 * oncoming barrage of data */
	pipe(nodeptr->err.fds);
	switch (fork()) {  /* its the ol fork and switch routine eh? */
	case -1:
	    bailout();
	    break;
	case 0: 
	    if (dup2(nodeptr->out.fds[1], STDOUT_FILENO) != STDOUT_FILENO)
		bailout();
	    if (dup2(nodeptr->err.fds[1], STDERR_FILENO) != STDERR_FILENO)
		bailout();
	    if (close(nodeptr->out.fds[0]) != 0)
		bailout();
	    if (close(nodeptr->err.fds[0]) != 0)
		bailout();
	    if (username != NULL)
		(void)snprintf(buf, MAXBUF, "%s@%s", username, nodeptr->name);
	    else
		(void)snprintf(buf, MAXBUF, "%s", nodeptr->name);
	    if (debug)
		    (void)printf("%s %s %s\n", rshstring, buf, command);
	    cmd = calloc(nrofargs+1, sizeof(char *));
	    arg = 0;
	    while (rsh[arg] != NULL) {
		    cmd[arg] = rsh[arg];
		    arg++;
	    }
	    cmd[arg++] = buf;
	    cmd[arg++] = command;
	    cmd[arg] = (char *)0;
	    execvp(rsh[0], cmd);
	    bailout();
	} /* end switch */
	/* now close off the useless stuff, and read the goodies */
	if (close(nodeptr->out.fds[1]) != 0)
	    bailout();
	if (close(nodeptr->err.fds[1]) != 0)
	    bailout();
	fda = fdopen(nodeptr->out.fds[0], "r"); /* stdout */
	if (fda == NULL)
	    bailout();
	fd = fdopen(nodeptr->err.fds[0], "r"); /* stderr */
	if (fd == NULL)
	    bailout();
	fds[0].fd = nodeptr->out.fds[0];
	fds[1].fd = nodeptr->err.fds[0];
	fds[0].events = POLLIN|POLLPRI;
	fds[1].events = POLLIN|POLLPRI;
	pollret = 1;
	while (pollret >= 0) {
	    int gotdata;

	    pollret = poll(fds, 2, 5);
	    gotdata = 0;
	    if ((fds[0].revents&POLLIN) == POLLIN ||
		(fds[0].revents&POLLHUP) == POLLHUP ||
		(fds[0].revents&POLLPRI) == POLLPRI) {
		cd = fgets(pipebuf, sizeof(pipebuf), fda);
		if (cd != NULL) {
		    (void)printf("%*s: %s",
				 -maxnodelen, nodeptr->name, cd);
		    gotdata++;
		}
	    }
	    if ((fds[1].revents&POLLIN) == POLLIN ||
		(fds[1].revents&POLLHUP) == POLLHUP ||
		(fds[1].revents&POLLPRI) == POLLPRI) {
		cd = fgets(pipebuf, sizeof(pipebuf), fd);
		if (errorflag && cd != NULL) {
		    (void)printf("%*s: %s",
				 -maxnodelen, nodeptr->name, cd);
		    gotdata++;
		} else if (!errorflag && cd != NULL)
		    gotdata++;
	    }
	    if (!gotdata)
		if (((fds[0].revents&POLLHUP) == POLLHUP ||
		     (fds[0].revents&POLLERR) == POLLERR ||
		     (fds[0].revents&POLLNVAL) == POLLNVAL) &&
		    ((fds[1].revents&POLLHUP) == POLLHUP ||
		     (fds[1].revents&POLLERR) == POLLERR ||
		     (fds[1].revents&POLLNVAL) == POLLNVAL))
		    break;
	}
	fclose(fda);
	fclose(fd);
	(void)wait(&status);
	if (piping) {
	    if (isatty(STDIN_FILENO) && piping)
	    /* yes, this is code repetition, no need to adjust your monitor */
		(void)printf("%s>", progname);
	    command = fgets(cbuf, sizeof(cbuf), in);
	    if (command != NULL)
		if (strcmp(command,"\n") == 0)
		    command = NULL;
	} else
	    command = NULL;
    } /* while loop */
    if (piping) {
	/* I learned this the hard way */
	fflush(in);
	fclose(in);
    }
}

void
sig_handler(int i)
{
    switch (i) {
    case SIGALRM:
	break;
    default:
	bailout();
	break;
    }
}
